// @holochain/conductor-api/lib/websocket/app.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppWebsocket = void 0;
/**
 * Defines AppWebsocket, an easy-to-use websocket implementation of the
 * Conductor API for apps
 *
 *    const client = AppWebsocket.connect(
 *      'ws://localhost:9000',
 *      signal => console.log('got a signal:', signal)
 *    )
 *
 *    client.callZome({...})  // TODO: show what's in here
 *      .then(() => {
 *        console.log('DNA successfully installed')
 *      })
 *      .catch(err => {
 *        console.error('problem installing DNA:', err)
 *      })
 */
const msgpack_1 = require("@msgpack/msgpack");
const client_1 = require("./client");
const common_1 = require("./common");
const common_2 = require("../api/common");
const launcher_1 = require("../environments/launcher");
class AppWebsocket {
    constructor(client, defaultTimeout, overrideInstalledAppId) {
        this.overrideInstalledAppId = overrideInstalledAppId;
        this._requester = (tag, transformer) => common_2.requesterTransformer((req, timeout) => common_1.promiseTimeout(this.client.request(req), tag, timeout || this.defaultTimeout).then(common_1.catchError), tag, transformer);
        this.appInfo = this._requester('app_info', appInfoTransform(this.overrideInstalledAppId));
        this.callZome = this._requester('zome_call_invocation', callZomeTransform);
        this.client = client;
        this.defaultTimeout = defaultTimeout === undefined ? common_1.DEFAULT_TIMEOUT : defaultTimeout;
    }
    static connect(url, defaultTimeout, signalCb) {
        return __awaiter(this, void 0, void 0, function* () {
            // Check if we are in the launcher's environment, and if so, redirect the url to connect to
            const env = yield launcher_1.getLauncherEnvironment();
            if (env) {
                url = `ws://localhost:${env.APP_INTERFACE_PORT}`;
            }
            const wsClient = yield client_1.WsClient.connect(url, signalCb);
            return new AppWebsocket(wsClient, defaultTimeout, env ? env.INSTALLED_APP_ID : undefined);
        });
    }
}
exports.AppWebsocket = AppWebsocket;
const callZomeTransform = {
    input: (req) => {
        return Object.assign(Object.assign({}, req), { payload: msgpack_1.encode(req.payload) });
    },
    output: (res) => {
        return msgpack_1.decode(res);
    },
};
const appInfoTransform = (overrideInstalledAppId) => ({
    input: (req) => {
        if (overrideInstalledAppId) {
            return {
                installed_app_id: overrideInstalledAppId,
            };
        }
        return req;
    },
    output: (res) => {
        return res;
    },
});
