// ton/dist/messages/wallet/WalletV3SigningMessage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WalletV3SigningMessage = void 0;
const Cell_1 = require("../../boc/Cell");
class WalletV3SigningMessage {
    timestamp;
    seqno;
    walletId;
    order;
    sendMode;
    constructor(args) {
        this.order = args.order;
        this.sendMode = args.sendMode;
        if (args.timestamp !== undefined && args.timestamp !== null) {
            this.timestamp = args.timestamp;
        }
        else {
            this.timestamp = Date.now();
        }
        if (args.seqno !== undefined && args.seqno !== null) {
            this.seqno = args.seqno;
        }
        else {
            this.seqno = 0;
        }
        if (args.walletId !== null && args.walletId !== undefined) {
            this.walletId = args.walletId;
        }
        else {
            this.walletId = 698983191;
        }
    }
    writeTo(cell) {
        cell.bits.writeUint(this.walletId, 32);
        if (this.seqno === 0) {
            for (let i = 0; i < 32; i++) {
                cell.bits.writeBit(1);
            }
        }
        else {
            const timestamp = Math.floor(this.timestamp / 1e3); // In seconds
            cell.bits.writeUint(timestamp + 60, 32); // 60 seconds timeout
        }
        cell.bits.writeUint(this.seqno, 32);
        cell.bits.writeUint8(this.sendMode);
        // Write order
        let orderCell = new Cell_1.Cell();
        this.order.writeTo(orderCell);
        cell.refs.push(orderCell);
    }
}
exports.WalletV3SigningMessage = WalletV3SigningMessage;
