// ton/dist/contracts/sources/WalletV2R1Source.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WalletV2R1Source = void 0;
const __1 = require("../..");
class WalletV2R1Source {
    static create(opts) {
        // Resolve parameters
        let publicKey = opts.publicKey;
        let workchain = opts.workchain;
        // Build initial code and data
        let initialCode = __1.Cell.fromBoc('B5EE9C724101010100570000AAFF0020DD2082014C97BA9730ED44D0D70B1FE0A4F2608308D71820D31FD31F01F823BBF263ED44D0D31FD3FFD15131BAF2A103F901541042F910F2A2F800029320D74A96D307D402FB00E8D1A4C8CB1FCBFFC9ED54A1370BB6')[0];
        let initialData = new __1.Cell();
        initialData.bits.writeUint(0, 32); // SeqNo
        initialData.bits.writeBuffer(publicKey); // Public key
        return new WalletV2R1Source({ publicKey, initialCode, initialData, workchain });
    }
    publicKey;
    initialCode;
    initialData;
    workchain;
    type = 'org.ton.wallets.v2';
    constructor(opts) {
        this.publicKey = opts.publicKey;
        this.initialCode = opts.initialCode;
        this.initialData = opts.initialData;
        this.workchain = opts.workchain;
        Object.freeze(this);
    }
}
exports.WalletV2R1Source = WalletV2R1Source;
